feat(jobs): add jobs for linkedlist rbegin/rend

danblitzhou opened the pull request:
- added jobs for `aws_linked_list_rbegin` and  `aws_linked_list_rend`
- added `ensure_linked_list_is_allocated_nd` in helper library; this function is useful for verifying linked list functions that only concern the front item and the back item (other than head and tail).

agurfinkel commented:
close, but lots of redundancies that raises some questions. Let's clean it up before accepting.

danblitzhou commented on seahorn/jobs/linked_list_rend/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+  MAX_BUFFER_SIZE=${MAX_BUFFER_SIZE}
+  MAX_LINKED_LIST_ITEM_ALLOCATION=${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE})
+sea_attach_bc(linked_list_rend)
+configure_file(sea.yaml sea.yaml @ONLY)
```
no job sea.yaml, will remove

agurfinkel replied:
also max size

danblitzhou commented on seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c:
```diff
@@ -0,0 +1,40 @@
+        aws_linked_list_push_back(&list, new_last);
+    }
+    /* Assume the preconditions */
+    assume(aws_linked_list_is_valid(&list));
```
It is if we keep `AWS_DEEP_CHECKS=0`; this essentially checks `list && list->head.next && list->head.prev == NULL && list->tail.prev && list->tail.next == NULL`; I believe I can make this compatible with the additional length > 2 models by adding `assume(list->head.next)` and `assume(list->tail.prev)`

agurfinkel replied:
maybe better to have our version to have full control of the pf

danblitzhou commented on seahorn/lib/linked_list_helper.c:
```diff
@@ -39,6 +38,37 @@ void ensure_linked_list_is_allocated(struct aws_linked_list *const list,
 }
 
+void ensure_linked_list_is_allocated_nd(struct aws_linked_list *const list,
+                                            size_t max_length, size_t *length) {
```
does `ensure_linked_list_is_initialized_nd` sound more reasonable?

agurfinkel replied:
not really. 

there are 3 different functions like this that we seem to need. The names should reflect that. Check Sidharth PR that is still open. 

I think we are using linked_front for forward linked list, linked_back for back linked.

danblitzhou commented:
addressed review comments:
- broke `ensure_linked_list_is_allocated_nd` into three function for different use case (specifically for length > 2)
- enabled `sea_aws_linked_list_is_valid` to check linked list shallowly, behaviour can be switched using macro constant `SEA_DEEP_CHECKS` similar to `AWS_DEEP_CHECKS`
- other renaming and small tweaks

danblitzhou commented on seahorn/jobs/linked_list_rend/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+  ${SEA_LIB}/proof_allocators.c
+  ${SEA_LIB}/linked_list_helper.c)
+target_compile_definitions(linked_list_rend
+  # We will use our own function for is_valid check hence AWS_DEEP_CHECK is off
```
promise is met!

agurfinkel commented on seahorn/jobs/linked_list_rbegin/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# add head and tail to list size.
+MATH(EXPR UNROLL_BOUND "${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE} + 2")
```
not needed

agurfinkel commented on seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c:
```diff
@@ -0,0 +1,40 @@
+    struct aws_linked_list list;
+    size_t length;
+    sea_nd_init_aws_linked_list_from_back(&list, &length);
+    // if (length > 2) {
```
remove

agurfinkel commented on seahorn/jobs/linked_list_rend/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# add head and tail to list size.
+MATH(EXPR UNROLL_BOUND "${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE} + 2")
```
not needed

agurfinkel reviewed (changes requested):
almost there

danblitzhou commented:
last review comments addressed
